sql/cagg_api.sql

```diff
@@ -0,0 +1,136 @@
+-- This file and its contents are licensed under the Apache License 2.0.
+-- Please see the included NOTICE for copyright information and
+-- LICENSE-APACHE for a copy of the license.
+
+-- Helper function to get information about a hypertable and the
+-- primary dimension.
+CREATE OR REPLACE FUNCTION _timescaledb_functions.get_hypertable_info(
+       hypertable REGCLASS,
+       column_type REGTYPE
+) RETURNS integer AS
+$body$
+DECLARE
+   info RECORD;
+BEGIN
+   SELECT ht.id AS hypertable_id, di.column_type::regtype INTO info
+     FROM hypertable ht
+     JOIN dimension di ON ht.id = di.hypertable_id
+    WHERE format('%I.%I', schema_name, table_name)::regclass = hypertable
+      AND di.interval_length IS NOT NULL;
+
+   IF info IS NULL THEN
+      RAISE 'table "%" is not a hypertable', hypertable;
+   END IF;
+
+   IF info.column_type <> get_hypertable_info.column_type THEN
+      RAISE 'wrong column type for hypertable %', hypertable
+      USING HINT = format('hypertable type was "%s", but caller expected "%s"',
+      	    	   	  info.column_type, get_hypertable_info.column_type);
+   END IF;
+
+   RETURN info.hypertable_id;
+END
+$body$ LANGUAGE plpgsql
+SET search_path = _timescaledb_catalog, _timescaledb_functions, public;
+
+-- Get hypertable invalidations ranges based on bucket size.
+--
+-- This will return a multirange for each bucket size passed in and a
+-- token that can be used to accept the multirange.
+--
+-- Note that the token returned is not unique for each bucket size and
+-- represents either the LSN or a Snapshot of what data was read to
+-- produce the bucket ranges.
+--
+-- Currently, we only have support for timestamp with and without
+-- timezone, but it is straightforward to add similar implementations
+-- for integer types.
+CREATE OR REPLACE FUNCTION _timescaledb_functions.get_hypertable_invalidations(
+   hypertable REGCLASS,
+   base TIMESTAMPTZ,
+   bucket_widths INTERVAL[]
+) RETURNS TABLE (bucket_width INTERVAL, token TEXT, invalidations TSTZMULTIRANGE) AS
```

Comment:
```suggestion
) RETURNS TABLE (bucket_width INTERVAL, token pg_snapshot, invalidations TSTZMULTIRANGE) AS
```
Can we use `pg_snapshot` datatype here?

Reply:
No, because once we start using the logical decoding, this will be a `pg_lsn` instead, so explicitly using `text` here.

Also, to be able to use the result of `pg_current_snapshot()`, we need to convert it to `text` before converting it to `xid8`.

Reply:
Is the plan to implement this same function interface for logical decoding?